### PR TITLE
[Feature] Add navigation witness

### PR DESCRIPTION
**Pull request by danyf90**
**Why**
Expose the navigation as a protocol witness to facilitate testing of the app logic

**Changes**
Add NavigationWitness struct

**Tasks**
* [X] Add relevant tests, if needed
* [ ] Add documentation, if needed
* [ ] Update README, if needed
* [ ] Ensure that the demo project works properly

### CODE REVIEW

File: Podfile.lock [hunk 1]
```diff
@@ -15,20 +15,17 @@ DEPENDENCIES:
   - Nimble (~> 7.3)
   - PinLayout
   - Quick (~> 1.3)
-  - Tempura (from `.`)
+  - Tempura (~> 6.0)
```
Comment (LorDisturbia) [inline Podfile.lock:18]: [i] Maybe let's have TempuraTesting point to the local Tempura in the Podfile, and check that everything is set up correctly in the podspec

File: Tempura/Navigation/NavigationWitness.swift [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+          }
+      }
+    )
+  }
+
+  public func installRoot(_ rootViewController: UIViewController) {
+    self.window.rootViewController = rootViewController
+    self.window.makeKeyAndVisible()
+  }
+}
```
Comment (LorDisturbia) [inline Tempura/Navigation/NavigationWitness.swift:192]: 🔥

File: Podfile.lock (hunk 1 above)
Comment (danyf90) [reply Podfile.lock]: ✅

File: Tempura/Navigation/NavigationWitness.swift [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    guard let identifierToHide = UIApplication.shared.currentRoutableIdentifiers.last else {
+      fatalError("No routable identifiers found")
+    }
+    return self.hide(identifierToHide, animated, context, atomic)
+  }
+}
+
+// MARK: - Live
+
+extension NavigationWitness {
```
Comment (FEC-bendingspoons) [inline Tempura/Navigation/NavigationWitness.swift:92]: why are we not wrapping this in a `#if DEBUG`? this stuff should NEVER reach prod

File: Tempura/Navigation/NavigationWitness.swift [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+      show: show,
+      hide: hide
+    )
+  }
+}
+
+// MARK: - Wrapped
+
+/// Wraps any object into a reference-type object
+public class Wrapped<T> {
```
Comment (FEC-bendingspoons) [inline Tempura/Navigation/NavigationWitness.swift:176]: [i] If you look in Ramen there is a `Box` type that does this but it also has dynamic dispatch to the underlying value (which is kinda cool to have). I'd put that here

**Review by FEC-bendingspoons — CHANGES_REQUESTED** [review]

File: Tempura/Navigation/NavigationWitness.swift (hunk 3 above)
Comment (LorDisturbia) [reply Tempura/Navigation/NavigationWitness.swift:92]: The live one needs to, actually. But I agree about the `mocked`, the `spy` and the `unimplemented`

File: Tempura/Navigation/NavigationWitness.swift (hunk 4 above)
Comment (danyf90) [reply Tempura/Navigation/NavigationWitness.swift]: Replaced `Wrapped` with ad hoc `NavigationRequests` type for improved ergonomic

**Review by FEC-bendingspoons — APPROVED** [review]

**Review by LorDisturbia — APPROVED** [review]

**Pull request merged**